hikari/components.py

```diff
@@ -0,0 +1,292 @@
+
+from __future__ import annotations
+
+__all__: typing.Sequence[str] = (
+    "ComponentType",
+    "PartialComponent",
+    "ActionRowComponent",
+    "ButtonStyle",
+    "ButtonComponent",
+    "SelectMenuOption",
+    "SelectMenuComponent",
+    "TextInputStyle",
+    "TextInputComponent",
+    "InteractiveButtonTypes",
+    "InteractiveButtonTypesT",
+)
+
+import typing
+
+import attr
+
+from hikari import emojis
+from hikari.internal import enums
+
+
+@typing.final
+class ComponentType(int, enums.Enum):
+    """Types of components found within Discord."""
+
+    ACTION_ROW = 1
+    """A non-interactive container component for other types of components.
+
+    !!! note
+        As this is a container component it can never be contained within another
+        component and therefore will always be top-level.
+
+    !!! note
+        As of writing this can only contain one component type.
+    """
+
+    BUTTON = 2
+    """A button component.
+
+    !!! note
+        This cannot be top-level and must be within a container component such
+        as `ComponentType.ACTION_ROW`.
+    """
+
+    SELECT_MENU = 3
+    """A select menu component.
+
+    !!! note
+        This cannot be top-level and must be within a container component such
+        as `ComponentType.ACTION_ROW`.
+    """
+
+    TEXT_INPUT = 4
+    """A text input component.
+
+    !! note
+        This component may only be used inside a modal container.
+
+    !!! note
+        This cannot be top-level and must be within a container component such
+        as `ComponentType.ACTION_ROW`.
+    """
+
+
+@typing.final
+class ButtonStyle(int, enums.Enum):
+    """Enum of the available button styles.
+
+    More information, such as how these look, can be found at
+    https://discord.com/developers/docs/interactions/message-components#buttons-button-styles
+    """
+
+    PRIMARY = 1
+    """A blurple "call to action" button."""
+
+    SECONDARY = 2
+    """A grey neutral button."""
+
+    SUCCESS = 3
+    """A green button."""
+
+    DANGER = 4
+    """A red button (usually indicates a destructive action)."""
+
+    LINK = 5
+    """A grey button which navigates to a URL.
+
+    !!! warning
+        Unlike the other button styles, clicking this one will not trigger an
+        interaction and custom_id shouldn't be included for this style.
+    """
+
+
+@typing.final
+class TextInputStyle(int, enums.Enum):
+    """A text input style."""
+
+    SHORT = 1
+    """Intended for short single-line text."""
+
+    PARAGRAPH = 2
+    """Intended for much longer inputs."""
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class PartialComponent:
+    """Base class for all component entities."""
+
+    type: typing.Union[ComponentType, int] = attr.field()
+    """The type of component this is."""
+
+
+@attr.define(weakref_slot=False)
+class ActionRowComponent(PartialComponent):
+    """Represents a row of components."""
+
+    components: typing.Sequence[PartialComponent] = attr.field()
```

Comment:
Using the same action row component object for both modals and message components is still annoying as far as type assurance goes, means you can't type something as taking specifically a message or modal action row as a library and expect to only get the sub-components for that specific type and also means that Hikari as a library couldn't down the line switch to a more narrow explicit type for components which only allows the types it'll get in the real world for the context it'll be in (e.g. a shared union of the message sub-component types for message components and a shared union of the modal sub-component types for modal sub components)

Reply:
I'm all for making actionrow generic but please not separate types anything but that.

surely something like this is enough right
```py
MessageActionRow = ActionRow[Button | SelectMenu]
ModalActionRow = ActionRow[TextInput | SelectMenu]
```

Reply:
Updated, please have a look at the latest commit